src/Database/LSMTree/Internal/IndexCompact.hs

```diff
@@ -436,42 +436,67 @@ pageSpanSize :: PageSpan -> NumPages
+-- there is no such pair, the result is an arbitrary but valid page span.
 --
 -- See [an informal description of the search algorithm](#search-descr) for more
 -- details about the search algorithm.
---
--- TODO: optimisation ideas: we should be able to get it all unboxed,
--- non-allocating and without unnecessary bounds checking.
---
+search :: SerialisedKey -> IndexCompact -> PageSpan
 -- One way to think of the search algorithm is that it starts with the full page
 -- number interval, and shrinks it to a minimal interval that contains the
--- search key. The code below is annotated with @Pre:@ and @Post:@ comments that
--- describe the interval at that point.
-search :: SerialisedKey -> IndexCompact -> PageSpan
+-- search key. The code below is annotated with [x,y] or [x, y) comments that
+-- describe the known page number interval at that point in the search
+-- algorithm.
 search k IndexCompact{..} =
-    let !primbits = keyTopBits64 k
-    in
-      case unsafeSearchLE primbits icPrimary of
-        Nothing -> singlePage (PageNo 0)  -- Post: @[lb, lb)@ (empty).
-        Just !i ->         -- Post: @[lb, i]@.
-          if unBit $ icClashes VU.! i then
-            -- Post: @[lb, i]@, now in clash recovery mode.
-            let !i1  = PageNo $ fromMaybe 0 $
-                  bitIndexFromToRev (BoundInclusive 0) (BoundInclusive i) (Bit False) icClashes
-                !i2  = maybe (PageNo 0) snd $ Map.lookupLE (unsafeNoAssertMakeUnslicedKey k) icTieBreaker
-                PageNo !i3 = max i1 i2 -- Post: the intersection of @[i1, i]@ and @[i2, i].
-                !i4 = bitLongestPrefixFromTo (BoundExclusive i3) (BoundInclusive i) (Bit True) icLargerThanPage
-                      -- Post: [i3, i4]
-            in  if i3 == i4 then singlePage (PageNo i3) else multiPage (PageNo i3) (PageNo i4)
-                -- Post: @[i3, i4]@ if a larger-than-page value that starts at
-                -- @i3@ and ends at @i4@, @[i3, i3]@ otherwise for a "normal"
-                -- page.
-          else  singlePage (PageNo i) -- Post: @[i, i]@
+    let !primbits = keyTopBits64 k in
+    -- [0, n), where n is the length of the P array
+    case unsafeSearchLE primbits icPrimary of
+      Nothing ->
+        -- TODO: if the P array is indeed empty, then this violates the
+        -- guarantee that we return a valid page span! We should specify that a
+        -- compact index should be non-empty.
+        if VU.length icLargerThanPage == 0 then singlePage (PageNo 0) else
+        -- [0, n), our page span definitely starts at 0, but we still have to
+        -- consult the LTP array to check whether the value on page 0 overflows
+        -- into subsequent pages.
+        let !i = bitLongestPrefixFromTo (BoundExclusive 0) NoBound (Bit True) icLargerThanPage
+        -- [0, i]
+        in  if i == 0
+            then singlePage (PageNo 0)
+                 -- [0, 0]
+            else multiPage (PageNo 0) (PageNo i)
+                 -- [0, i]
```

Comment:
Isn't this the same as https://github.com/IntersectMBO/lsm-tree/pull/401#discussion_r1776696453, so we could remove the check?

Reply:
Definitely!